spinoso-time/src/time/tzrs/offset.rs

```diff
@@ -0,0 +1,273 @@
+}
+
+/// Represents the number of seconds offset from UTC
+#[allow(variant_size_differences)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Offset {
+    /// UTC offset, zero offset, Zulu time
+    Utc,
+    /// Fixed offset from UTC
+    ///
+    /// Note: A fixed offset of 0 is different from UTC time
+    Fixed([LocalTimeType; 1]),
+    /// A time zone based offset
+    Tz(TimeZoneRef<'static>),
+}
+
+impl<'a> Offset {
+    /// Generate a UTC based offset
+    #[inline]
+    #[must_use]
+    pub fn utc() -> Self {
+        Self::Utc
+    }
+
+    /// Generate an offset based on the detected local time zone of the system
+    ///
+    /// Detection is done by [`tzdb::local_tz`], and if it fails will return a GMT timezone
+    ///
+    /// [`tzdb::local_tz`]: https://docs.rs/tzdb/latest/tzdb/fn.local_tz.html
+    #[inline]
+    #[must_use]
+    pub fn local() -> Self {
+        Self::Tz(local_time_zone())
+    }
+
+    /// Generate an offset with a number of seconds from UTC.
+    #[inline]
+    #[must_use]
+    pub fn fixed(offset: i32) -> Self {
+        let offset_name = offset_hhmm_from_seconds(offset);
+        let local_time_type =
+            LocalTimeType::new(offset, false, Some(offset_name.as_bytes())).expect("Couldn't create fixed offset");
+        Self::Fixed([local_time_type])
+    }
+
+    /// Generate an offset based on a provided [`tz::timezone::TimeZoneRef`]
+    ///
+    /// This can be combined with [`tzdb`] to generate offsets based on predefined iana time zones
+    ///
+    /// ```
+    /// use spinoso_time::tzrs::Offset;
+    /// use tzdb::time_zone::pacific::AUCKLAND;
+    /// let offset = Offset::tz(AUCKLAND);
+    /// ```
+    ///
+    /// [`tz:timezone::TimeZoneRef`]: https://docs.rs/tz-rs/0.6.9/tz/timezone/struct.TimeZoneRef.html
+    /// [`tzdb`]: https://docs.rs/tzdb/latest/tzdb/index.html
+    #[inline]
+    #[must_use]
+    pub fn tz(tz: TimeZoneRef<'static>) -> Self {
+        Self::Tz(tz)
+    }
+
+    /// Returns a `TimeZoneRef` which can be used to generate and project _time_.
+    #[inline]
+    #[must_use]
+    pub fn time_zone_ref(&'a self) -> TimeZoneRef<'a> {
+        match self {
+            Self::Utc => TimeZoneRef::utc(),
+            Self::Fixed(local_time_types) => match TimeZoneRef::new(&[], local_time_types, &[], &None) {
+                Ok(tz) => tz,
+                Err(_) => GMT,
+            },
+
+            Self::Tz(zone) => *zone,
+        }
+    }
+}
+
+impl From<&str> for Offset {
+    /// Construct a Offset based on the [accepted MRI values]
+    ///
+    /// Accepts:
+    ///
+    /// - `[+/-]HH[:]MM`
+    /// - A-I representing +01:00 to +09:00
+    /// - K-M representing +10:00 to +12:00
+    /// - N-Y representing -01:00 to -12:00
+    /// - Z representing 0 offset
+    ///
+    /// [accepted MRI values]: https://ruby-doc.org/core-2.6.3/Time.html#method-c-new
+    #[inline]
+    #[must_use]
+    fn from(input: &str) -> Self {
```

Comment:
woohoo!

Reply:
Just wait until you realise that `Self::fixed(1)` gives a 1 second offset, not 1 hour offset. Just discovered this when attempting the `artichoke-backend` integration :grimacing:. I'll fix it here and fix the merge conflicts up the chain though :smile: